### PR TITLE
Add expression functions for arrays and maps

### PR DESCRIPTION
**Pull request by pvalsecc**

### CODE REVIEW

**Comment by nyalldawson** [comment]
Awesome - this makes me very excited!

File: resources/function_help/json/array_concat [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "array_concat",
```
Comment (nyalldawson) [inline resources/function_help/json/array_concat]: This is differing from Postgres' naming ("array_cat"). Not sure if that's an issue or not - personally I prefer your "array_concat".

File: src/core/qgsexpression.cpp [hunk 2]
```diff
@@ -3156,6 +3156,143 @@ static QVariant fcnGetLayerProperty( const QVariantList& values, const QgsExpres
...
 }
 
+static QVariant fcnArray( const QVariantList& values, const QgsExpressionContext*, QgsExpression* )
+{
+  return values;
+}
+
+static QVariant fcnArrayLength( const QVariantList& values, const QgsExpressionContext*, QgsExpression* )
+{
+  return values.at( 0 ).toList().length();
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: It'd be a good idea to protect these conversions with a similar function to the existing getIntValue/getDoubleValue/etc, and set the eval error string to notify users if they use a non-list value with these functions.

File: src/core/qgsexpression.cpp [hunk 3]
```diff
@@ -3156,6 +3156,143 @@ static QVariant fcnGetLayerProperty( const QVariantList& values, const QgsExpres
...
+
+static QVariant fcnArrayFind( const QVariantList& values, const QgsExpressionContext*, QgsExpression* )
+{
+  return values.at( 0 ).toList().indexOf( values.at( 1 ) );
+}
+
+static QVariant fcnArrayGet( const QVariantList& values, const QgsExpressionContext*, QgsExpression* )
+{
+  const QVariantList list = values.at( 0 ).toList();
+  const int pos = values.at( 1 ).toInt();
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: these should use getIntValue

File: tests/src/core/testqgsexpression.cpp [hunk 4]
```diff
@@ -2148,6 +2148,173 @@ class TestQgsExpression: public QObject
...
+      QCOMPARE( QgsExpression( "array_remove_at(\"strings\", 0)" ).evaluate( &context ), QVariant( removeAtExpected ) );
+
+      QStringList removeAllExpected;
+      removeAllExpected << "a" << "b" << "d";
+      QCOMPARE( QgsExpression( "array_remove_all(array('a', 'b', 'c', 'd', 'c'), 'c')" ).evaluate( &context ), QVariant( removeAllExpected ) );
+
+      QStringList concatExpected = array;
+      concatExpected << "a" << "b" << "c";
+      QCOMPARE( QgsExpression( "array_concat(\"strings\", array('a', 'b'), array('c'))" ).evaluate( &context ), QVariant( concatExpected ) );
+    }
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp]: I'd like to see some tests here where the array/map functions are called with non array/map values

**Review by nyalldawson — CHANGES_REQUESTED** [review]
Very very comprehensive PR - great work! Two things which I was thinking after looking over this:
- what do you think about adding an "array_intersects(a,b)" function which would return true if an any element of a exists in b?
- do you think there's any value in having a way of writing an array literal, eg "[2,3,4]"? I'm happy to have a go at this (I really enjoy working with the lexer now!) unless you think it adds no value over the `array(a,b)` function. 

This looks fantastic... it opens up stacks of possibilities. I'm already thinking an array aggregate function, array to multigeom, split string to array, array to string (with concatenator),...

File: resources/function_help/json/array_concat (hunk 1 above)
Comment (pvalsecc) [inline resources/function_help/json/array_concat]: I've been asked to match the postgres function names. Mistake on my side. I'll change it.

**Comment by nyalldawson** [comment]
Another cool thing we could do:

` apply(@element * 2, array(1,2,3) ) ` ->2, 4, 6

**Comment by nirvn** [comment]
@pvalsecc , nice!

It'd be useful to be able to create arrays and maps out of a string; that way, an array or map could be created out of a string field.

For e.g, if you open an .osm (openstreetmap file format) in QGIS, there's an other_tag field which adds additional osm tags that aren't mapped out by ogr using a string: 
"access"=>"yes","bicycle"=>"yes","cycleway"=>"no","foot"=>"yes","maxspeed"=>"30","oneway"=>"no","surface"=>"asphalt"

We could then do stuff like: map_get(map("other_tags"),oneway)

**Comment by nyalldawson** [comment]
(thinking aloud) another useful function would be `range(min,max,step),` which returns an array

**Comment by pvalsecc** [comment]
OK, I've fixed all the comments but the ones asking for additional functions (except array_intersect).

For the other functions, they can be added later. Otherwise this PR will just keep growing with people's requests.

For the string parsing, there are so many formats out there to represent maps and arrays, that will require a good reflection before starting.

**Comment by nyalldawson** [comment]
Agreed about not adding extra functions here - the list would be endless ;)

**Comment by pvalsecc** [comment]
@nirvn,

Wouldn't it be better to change the stuff that reads .osm to import the "other_tags" column as a QVariantMap? Like I did for postgres and spatialite?

**Comment by nirvn** [comment]
@pvalsecc , sure, it'd be fantastic if that's possible.

**Comment by rouault** [comment]
> Wouldn't it be better to change the stuff that reads .osm to import the "other_tags" column as a QVariantMap? Like I did for postgres and spatialite?

FYI, the "stuff" in that instance if the OGR OSM driver ;-). There's no string map type currently in OGR. Hence this serialization as a Postgres HSTORE

File: src/core/qgsexpression.cpp [hunk 5]
```diff
@@ -3491,6 +3661,30 @@ const QList<QgsExpression::Function*>& QgsExpression::Functions()
...
+    //functions for maps
+    << new StaticFunction( "map", -1, fcnMap, "Maps" )
+    << new StaticFunction( "map_get", ParameterList() << Parameter( "map" ) << Parameter( "key" ), fcnMapGet, "Maps" )
+    << new StaticFunction( "map_exist", ParameterList() << Parameter( "map" ) << Parameter( "key" ), fcnMapExist, "Maps" )
+    << new StaticFunction( "map_delete", ParameterList() << Parameter( "map" ) << Parameter( "key" ), fcnMapDelete, "Maps" )
+    << new StaticFunction( "map_insert", ParameterList() << Parameter( "map" ) << Parameter( "key" ) << Parameter( "value" ), fcnMapInsert, "Maps" )
+    << new StaticFunction( "map_concat", -1, fcnMapConcat, "Maps" )
+    << new StaticFunction( "map_akeys", ParameterList() << Parameter( "map" ), fcnMapAKeys, "Maps" )
+    << new StaticFunction( "map_avals", ParameterList() << Parameter( "map" ), fcnMapAVals, "Maps" )
     ;
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: These new functions must also be added to QgsExpression::BuiltinFunctions

**Comment by nyalldawson** [comment]
Sorry - one last change noted above

File: src/core/qgsexpression.cpp (hunk 5 above)
Comment (pvalsecc) [inline src/core/qgsexpression.cpp]: Can't this list be built in this function based on the `StaticFunction`s we add in there?
I'll have a look Thursday. 3 days weekend, here.

**Comment by pvalsecc** [comment]
@nyalldawson, I've gone a bit freestyle for fixing the `BuiltinFunctions` problem.
Fixes the missing functions in the list.

**Comment by nyalldawson** [comment]
Looks great! Any objections if I backport these 2 commits to master_2? (just the expression stuff, not the provider or widget changes)

**Comment by pvalsecc** [comment]
> Any objections if I backport these 2 commits to master_2? (just the expression stuff, not the provider or widget changes)

No problem backport whatever you see fit.

**Pull request merged**